### PR TITLE
Simple nest/frequent-spawn atlas

### PR DESCRIPTION
**Pull request by pki0**
## Description
- Add a simple nest atlas base on spawns in the last 24 hours
- Map with spawns > 25% (>6 times) a day per spawnpoint
- Map is loaded by .json data to speed it up
- Nest data is updated into json file every 15 minutes by the cronjob
- Thanks to @Arengo for the nest-pokemon_id's
- Should work with forced: "EN", "DE", "" (nothing)

## Fixes (WIP)
- ONLY_FULL_GROUP_BY error fixed. (27.03.2017)
- Add some Pokemon to nest-list (27.03.2017)

## QUERY
- New query:
`SELECT p.pokemon_id, max(p.latitude) as latitude, max(p.longitude) as longitude, count(p.pokemon_id) as total_pokemon FROM pokemon p INNER JOIN spawnpoint s ON (p.spawnpoint_id = s.id) WHERE p.disappear_time > UTC_TIMESTAMP - INTERVAL 24 HOUR AND p.pokemon_id IN(1, 4, 7, 23, 25, 27, 29, 32, 37, 43, 50, 56, 58, 63, 66, 69, 74, 77, 79, 81, 84, 86, 90, 92, 95, 98, 100, 102, 104, 108, 109, 111, 114, 116, 120, 123, 124, 125, 126, 127, 138, 140, 152, 155, 158, 170, 183, 185, 187, 190, 191, 193, 194, 200, 202, 203, 204, 206, 207, 209, 211, 213, 215, 216, 218, 223, 226, 227, 228, 231, 234, 241) GROUP BY p.spawnpoint_id, p.pokemon_id HAVING count(p.pokemon_id) >= 6 ORDER BY p.pokemon_id`

- Old query:
<s>`SELECT pokemon.pokemon_id, count(pokemon.pokemon_id) as total_pokemon, pokemon.latitude, pokemon.longitude FROM pokemon INNER JOIN spawnpoint ON (spawnpoint.id = pokemon.spawnpoint_id) WHERE pokemon.disappear_time > UTC_TIMESTAMP - INTERVAL 24 HOUR AND pokemon.pokemon_id IN(1, 4, 7, 23, 25, 27, 29, 32, 37, 43, 50, 56, 58, 60, 63, 66, 69, 74, 77, 79, 81, 84, 86, 90, 92, 95, 98, 100, 102, 104, 108, 109, 111, 114, 116, 120, 123, 124, 125, 126, 127, 138, 140, 152, 155, 158, 170, 183, 185, 187, 190, 191, 193, 194, 200, 202, 203, 204, 206, 207, 209, 211, 213, 215, 216, 218, 223, 226, 227, 228, 231, 234, 241) GROUP BY pokemon.spawnpoint_id, pokemon.pokemon_id HAVING total_pokemon > 6 ORDER BY pokemon.pokemon_id`</s>

## How Has This Been Tested?
- Locally with different forced languages
https://pokemonkiel.de/nests

## Screenshots (if appropriate):
image

## Types of changes
- [x] New feature (non-breaking change which adds functionality)

### CODE REVIEW

**Comment by mspitzbarth** [comment]
If sql_mode=only_full_group_by is activated, your code inside nests.cron.php doesnt´t work.

Error: Expression #3 of SELECT list is not in GROUP BY clause and contains nonaggregated column 'DBNAME.pokemon.latitude' which is not functionally dependent on columns in GROUP BY clause; this is incompatible with sql_mode=only_full_group_by

So i have to do 
mysql > SET GLOBAL sql_mode=(SELECT REPLACE(@@sql_mode,'ONLY_FULL_GROUP_BY',''));

**Comment by mpw1337** [comment]
Eevee (133) can also have nests.

**Comment by pki0** [comment]
@mpw1337 Are your sure? I did not found anything about that

@Neayto Thanks for the hint! I don't know if I can set this via php

**Comment by pki0** [comment]
@Neayto Error sould by fixed now. I have a result even with 

`SET sql_mode = 'ONLY_FULL_GROUP_BY';  `

`SELECT @@sql_mode;

+--------------------+

| @@sql_mode         |

+--------------------+

| ONLY_FULL_GROUP_BY |
+--------------------+`

Would be great if you can test the new query. Thanks!

`SELECT p.pokemon_id, max(p.latitude) as latitude, max(p.longitude) as longitude, count(p.pokemon_id) as total_pokemon FROM pokemon p INNER JOIN spawnpoint s ON (p.spawnpoint_id = s.id) WHERE p.disappear_time > UTC_TIMESTAMP - INTERVAL 24 HOUR AND p.pokemon_id IN(1, 4, 7, 23, 25, 27, 29, 32, 37, 43, 50, 56, 58, 63, 66, 69, 74, 77, 79, 81, 84, 86, 90, 92, 95, 98, 100, 102, 104, 108, 109, 111, 114, 116, 120, 123, 124, 125, 126, 127, 138, 140, 152, 155, 158, 170, 183, 185, 187, 190, 191, 193, 194, 200, 202, 203, 204, 206, 207, 209, 211, 213, 215, 216, 218, 223, 226, 227, 228, 231, 234, 241) GROUP BY p.spawnpoint_id, p.pokemon_id HAVING count(p.pokemon_id) >= 6 ORDER BY p.pokemon_id`

**Comment by mpw1337** [comment]
@pokemonkiel Yes, Eevee is definitely a nest pokemon. Even Spearow, Zubat and Sentret can have nests.
Current Eevee nest in our Park:
image
Link to Silphroad: [https://thesilphroad.com/atlas#16/49.4632/8.1776](https://thesilphroad.com/atlas#16/49.4632/8.1776)
Or Luisenpark in Mannheim Nest 
[SILPH NEST: #92573, 12th Migration](https://thesilphroad.com/atlas#16/49.4822/8.5022)

**Comment by pki0** [comment]
@mpw1337 Thanks. I'll add them now!

**Comment by mpw1337** [comment]
@pokemonkiel actually i think nearly every species, except dratini, mareep, larvitar, unown and evolved ones can have nests. But I don't have any proof atm.

**Comment by pki0** [comment]
@mpw1337 No I think Pokemon like Pidgey, Weedle, Rattata and very common urban biomes can't be a nest at all. So I'm wondering about Spearow

**Comment by mpw1337** [comment]
@pokemonkiel 
You can take a look yourself here: http://pkmnstat.amsel-ap.de/pokemon/21 nest mitigation Nr.12 
Spearow Nest:
image
Before the nest mitigation to Spearow:
image
and for Caterpie, Nest mitigation Nb. 10 http://pkmnstat.amsel-ap.de/pokemon/10 just compare it to the mitigation before.

**Comment by Obihoernchen** [comment]
Merge ETA: next week. Sorry for the delay.

**Comment by pki0** [comment]
@Obihoernchen No hurry :)

We have to discuss this nest thing a bit more:

@mpw1337 I think I got it. You can't figure out this as a Nest in my opinion! (Too few points on Heatmap)
Your Spearow "Nest" is in a non-nesting area so it can't be a nest. 

bildschirmfoto 2017-03-28 um 09 17 45

I have the same with Caterpie. But this is a non-nesting area, so no nest at all for me.

It's some kind of biome spawns. Maybe the area is more not so much mixed.

Have a look here:
bildschirmfoto 2017-03-28 um 09 25 29

I can't say this is a Dragonite Nest. It's just the biome thing.

Other good example is Phanpy:
bildschirmfoto 2017-03-28 um 09 44 30

Phanpy  seems to have the same biome as Dragonite. But this is not a Nest at all.

But I'll try to clear that on reddit, maybe I can get an answer. 👍 

@Arengo Where did you get the nesting Pokemon List?

**Comment by mpw1337** [comment]
@pokemonkiel this "Kurpark" is definitely a nesting area.
Take a look at Mitigation nb 10 or 11 with Rhyhorn 
And gen2 Mitigation (16.02.2017) Growlithe
That's not because of a biome.

**Comment by mspitzbarth** [comment]
@pokemonkiel is working fine now 👍

**Comment by mean-jive** [comment]
As far as I know, most 'mons can nest, like mpw1337 said; only a select few can't. The reason people don't think of pidgey etc nests, is cause you don't react to them when you encounter more than 1.

**Comment by mspitzbarth** [comment]
@pokemonkiel You should be able to define the pokemon id´s for the nest and the total_pokemons  inside variables.json. So you can change it to your settings what you like.

And use it inside nests.cron.php
When i am at home i can look through it :)

**Comment by pki0** [comment]
@Neayto Thanks. Glad to see that query is working now.
I added to variables string to decide pokemon_ids to show. Great idea. Thanks!

If someone has the trick to use new lines in variables.json tell me :)

**Comment by z0n** [comment]
I'm pretty sure that Swinub (#<span>220</span>) is also a nest Pokémon so it should be added to the default nest Pokémon list.
Before: image

After: image

**Comment by mean-jive** [comment]
https://www.reddit.com/r/TheSilphRoad/comments/62cxzz/the_hopefully_ultimate_list_of_nonnesting_species/ - As much of a complete list I could find.

**Review by Obihoernchen — CHANGES_REQUESTED** [review]
Super awesome stuff!
Please add lat,long to the infowindow.

**Comment by Obihoernchen** [comment]
We should exclude some very common pokemon like 21 and 161 from default nest_pokemon because nobody cares. They just clutter the map.

File: core/json/variables.examples.json [hunk 1]
```diff
@@ -18,21 +18,22 @@
...
-		"pokeimg_suffix"	:	"_flat.png",
-		"recents_show_iv"	:	true,
-		"recents_filter"	:	true,
+		"map_center_lat"		:	"50.844441",
+		"map_center_long"		:	"4.363557",
+		"zoom_level"			:	"13",
+		"GMaps_Key"				:	"#GMAPS_KEY#",
+		"timezone"				:	"Europe/Paris",
+		"max_pokemon"			:	251,
+		"nest_pokemon"			:	"1, 4, 7, 21, 23, 25, 27, 29, 32, 37, 41, 43, 50, 56, 58, 60, 63, 66, 69, 74, 77, 79, 81, 84, 86, 90, 92, 95, 98, 100, 102, 104, 108, 109, 111, 114, 116, 120, 123, 124, 125, 126, 127, 133, 138, 140, 152, 155, 158, 161, 170, 183, 185, 187, 190, 191, 193, 194, 200, 202, 203, 204, 206, 207, 209, 211, 213, 215, 216, 218, 223, 226, 227, 228, 231, 234, 241",
```
Comment (Obihoernchen) [inline core/json/variables.examples.json]: use an array instead and join it with with , as separator for the sql query (btw I removed 21 and 161).
Then we can use line breaks as well:

```json
[1, 4, 7, 23, 25, 27, 29, 32, 37, 41, 43, 50, 56, 58, 60, 63, 66, 69, 74,
77, 79, 81, 84, 86, 90, 92, 95, 98, 100, 102, 104, 108, 109, 111, 114, 116,
120, 123, 124, 125, 126, 127, 133, 138, 140, 152, 155, 158, 170, 183,
185, 187, 190, 191, 193, 194, 200, 202, 203, 204, 206, 207, 209, 211, 213, 
215, 216, 218, 223, 226, 227, 228, 231, 234, 241],
```

**Review by Obihoernchen — CHANGES_REQUESTED** [review]

**Comment by Obihoernchen** [comment]
What's the reason for using max?
`max(p.latitude) as latitude, max(p.longitude) as longitude,`

**Review by Obihoernchen — APPROVED** [review]

**Comment by Obihoernchen** [comment]
I think it's fine for now!
One more +1 and I'll merge.

We can improve the pokemon_id list later on.

**Comment by pki0** [comment]
@Obihoernchen. Could you please change included Pokemon IDs to excluded Pokemon IDs like on my very first commit.

This makes variables.json more readable and overall handling easier.
Change query from IN to NOT IN 

I think this is the better way

**Comment by Obihoernchen** [comment]
@pokemonkiel I agree. Done :+1:

File: core/cron/nests.cron.php [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+INNER JOIN spawnpoint s ON (p.spawnpoint_id = s.id) 
+WHERE p.disappear_time > UTC_TIMESTAMP - INTERVAL 24 HOUR 
+AND p.pokemon_id NOT IN (".$nest_exclude_pokemon_ids.")
+GROUP BY p.spawnpoint_id, p.pokemon_id 
+HAVING count(p.pokemon_id) >= 6 
+ORDER BY p.pokemon_id";
+$result = $mysqli->query($req);
+
+while ($data = $result->fetch_object()) {
+	$pokeid = $data->pokemon_id;
```
Comment (friscoMad) [inline core/cron/nests.cron.php]: I think this is not needed

Comment (pki0) [reply core/cron/nests.cron.php]: What do you mean exactly?

Comment (friscoMad) [reply core/cron/nests.cron.php]: $pokeid var was not needed but @Obihoernchen did just fix that.

**Review by Obihoernchen — APPROVED** [review]

**Review by friscoMad — APPROVED** [review]

File: core/js/nests.maps.js.php [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+	header('Content-type: text/javascript');
+
+	# Load variables and locales
+	include_once('../../config.php');
+	$variables = SYS_PATH.'/core/json/variables.json';
+	$config = json_decode(file_get_contents($variables));
+	include_once('../process/locales.loader.php');
+
+	# Load nests-file
+	$nest_file = file_get_contents('../json/nests.stats.json');
```
Comment (mpw1337) [inline core/js/nests.maps.js.php:12]: @pokemonkiel shouldn't this be:
`$nest_file = file_get_contents(SYS_PATH.'/core/json/nests.stats.json');`

Comment (Obihoernchen) [reply core/js/nests.maps.js.php:12]: It doesn't matter

Comment (mpw1337) [reply core/js/nests.maps.js.php:12]: on my setup, which has limited htaccess rights it didn't work without the SYS_PATH

**Pull request merged**